pulp_smash/tests/pulp3/file/api_v3/test_download_content.py

```diff
@@ -0,0 +1,100 @@
+# coding=utf-8
+"""Tests that verify download of content served by Pulp."""
+import hashlib
+import unittest
+from random import choice
+from urllib.parse import urljoin
+
+from pulp_smash import api, config, utils
+from pulp_smash.constants import FILE_FEED_URL
+from pulp_smash.tests.pulp3.constants import (
+    DISTRIBUTION_PATH,
+    FILE_IMPORTER_PATH,
+    FILE_PUBLISHER_PATH,
+    REPO_PATH,
+)
+from pulp_smash.tests.pulp3.file.api_v3.utils import (
+    gen_importer,
+    gen_publisher,
+)
+from pulp_smash.tests.pulp3.file.utils import set_up_module as setUpModule  # noqa pylint:disable=unused-import
+from pulp_smash.tests.pulp3.pulpcore.utils import gen_distribution, gen_repo
+from pulp_smash.tests.pulp3.utils import (
+    get_auth,
+    get_name_content_units,
+    publish_repo,
+    sync_repo,
+    read_repo_content,
+)
+
+
+class DownloadContentTestCase(unittest.TestCase, utils.SmokeTest):
+    """Verify whether content served by pulp can be downloaded."""
+
+    def test_all(self):
+        """Verify whether content served by pulp can be downloaded.
+
+        This test explores the design choice stated in `Pulp #2895`_ that
+        content served by pulp can be downloaded.
+
+        A distribution serves a publication which was created from a
+        repository version by a publisher.
+
+        Furthermore a distribution defines at which relative path on Pulp's
+        web server the publication will be served.
+
+        The ``base_path`` of a distribution is appended to ``/content/`` to
+        form the full path that content will be served.
+
+        .. _Pulp #2895: https://pulp.plan.io/issues/2895
+
+        This test targets the following issue:
+
+        * `Pulp Smash #872 <https://github.com/PulpQE/pulp-smash/issues/872>`_
+
+        Do the following:
+
+        1. Create necessary resources to server content in Pulp.
+           Repository, importer, distribution, publisher and publication.
+        2. Read the checksum of content units present in the repository.
+        3. Randomly select a unit to be downloaded.
+        4. Assert that the checksum of the downloaded unit is equal to
+           checksum of the unit in repository.
+        """
+        cfg = config.get_config()
+        client = api.Client(cfg, api.echo_handler)
+        client.request_kwargs['auth'] = get_auth()
+        repo = client.post(REPO_PATH, gen_repo()).json()
+        self.addCleanup(client.delete, repo['_href'])
+        body = gen_importer()
+        body['feed_url'] = urljoin(FILE_FEED_URL, 'PULP_MANIFEST')
+        importer = client.post(FILE_IMPORTER_PATH, body).json()
+        self.addCleanup(client.delete, importer['_href'])
+        sync_repo(cfg, importer, repo)
+        content_units = read_repo_content(repo)['results']
+        units_checksum = {k: content_unit['digest']
+                          for content_unit in content_units
+                          for k, v in content_unit['artifacts'].items()}
```

Comment:
Since `v` isn't used, don't create the variable in the first place. Instead, iterate through `content_unit['artifacts']`. Also, please don't pair up set comprehensions in this manner. I understand it, but it makes my head hurt.